drivers/adc/ad796x/ad796x.c

```diff
@@ -0,0 +1,297 @@
+*******************************************************************************/
+
+/******************************************************************************/
+/***************************** Include Files **********************************/
+/******************************************************************************/
+#include <stdlib.h>
+#include <errno.h>
+#include "ad796x.h"
+#include "no_os_alloc.h"
+#include "no_os_print_log.h"
+
+/******************************************************************************/
+/********************** Macros and Constants Definitions **********************/
+/******************************************************************************/
+
+#define AD796X_BYTES_PER_SAMPLE		4
+#define AD796X_NUM_GPIOS		4
+
+/******************************************************************************/
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+
+struct ad796x_mode_gpios {
+	uint8_t en0;
+	uint8_t en1;
+	uint8_t en2;
+	uint8_t en3;
+};
+
+static struct ad796x_mode_gpios ad796x_modes[] = {
+	[AD796X_MODE1_EXT_REF_5P0] = 		{.en3 = 1, .en2 = 0, .en1 = 0, .en0 = 1},
+	[AD796X_MODE2_INT_REF_4P0] = 		{.en3 = 1, .en2 = 0, .en1 = 0, .en0 = 1},
+	[AD796X_MODE3_EXT_REF_4P0] = 		{.en3 = 1, .en2 = 0, .en1 = 1, .en0 = 0},
+	[AD796X_MODE4_SNOOZE] = 		{.en3 = 1, .en2 = 0, .en1 = 1, .en0 = 1},
+	[AD796X_MODE5_TEST] = 			{.en3 = 0, .en2 = 1, .en1 = 0, .en0 = 0},
+	[AD796X_MODE6_INVALID] = 		{.en3 = 1, .en2 = 1, .en1 = 0, .en0 = 0},
+	[AD796X_MODE7_EXT_REF_5P0_9MHZ] = 	{.en3 = 1, .en2 = 1, .en1 = 0, .en0 = 1},
+	[AD796X_MODE8_INT_REF_4P0_9MHZ] = 	{.en3 = 1, .en2 = 1, .en1 = 0, .en0 = 1},
+	[AD796X_MODE9_EXT_REF_4P0_9MHZ] = 	{.en3 = 1, .en2 = 1, .en1 = 1, .en0 = 0},
+	[AD796X_MODE10_SNOOZE2] = 		{.en3 = 1, .en2 = 1, .en1 = 1, .en0 = 1},
+};
+
+/**
+ * @brief Read data samples from adc
+ * @param dev - The device structure.
+ * @param buf - The buffer to fill.
+ * @param samples - number of samples to read.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int32_t ad796x_read_data(struct ad796x_dev *dev,
+			 uint32_t *buf,
+			 uint16_t samples)
+{
+	int ret;
+
+	struct 	axi_dma_transfer read_transfer = {
+		.size = samples * AD796X_BYTES_PER_SAMPLE,
+		.transfer_done = 0,
+		.cyclic = NO,
+		.src_addr = 0,
+		.dest_addr = (uintptr_t)buf,
+	};
+
+	ret = axi_dmac_transfer_start(dev->axi_dmac, &read_transfer);
+	if (ret) {
+		pr_err("axi_dmac_transfer_start() failed!\n");
+		return ret;
+	}
+
+	return axi_dmac_transfer_wait_completion(dev->axi_dmac, 3000);
+}
+
+/**
+ * @brief Remove the device gpio resources.
+ * @param dev - The device structure.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+static int ad796x_gpio_remove(struct ad796x_dev *dev)
+{
+	int ret, i;
+
+	struct no_os_gpio_desc *desc[] = {
+		dev->gpio_adc_en3_fmc,
```

Comment:
dev should be checked for null before trying to access its members

Reply:
ill use a table index instead to loop though the gpios.